Sources/TSCBasic/Downloader.swift

```diff
@@ -0,0 +1,136 @@
+    private var downloads: [Int: Download] = [:]
+
+    /// The operation queue used to synchronize access to the class properties.
+    private let queue = OperationQueue()
+
+    /// Creates a `FoundationDownloader` with an optional `URLSessionConfiguration`.
+    ///
+    /// - Parameters:
+    ///   - configuration: The `URLSessionConfiguration` to setup the `URLSession` used for downloading. If left out, it
+    ///   will default to a `default` configuration.
+    public init(configuration: URLSessionConfiguration = .default) {
+        super.init()
+
+        queue.name = "org.swift.swiftpm.basic.foundation-downloader"
+        session = URLSession(configuration: configuration, delegate: self, delegateQueue: queue)
+    }
+
+    public func downloadFile(
+        at url: URL,
+        progress: @escaping (Double) -> Void,
+        completion: @escaping (Result<AbsolutePath, DownloaderError>) -> Void
+    ) {
+        queue.addOperation {
+            let task = self.session.downloadTask(with: url)
+            let download = Download(task: task, progress: progress, completion: completion)
+            self.downloads[task.taskIdentifier] = download
+            task.resume()
+        }
+    }
+}
+
+extension FoundationDownloader: URLSessionDownloadDelegate {
+    public func urlSession(
+        _ session: URLSession,
+        downloadTask: URLSessionDownloadTask,
+        didWriteData bytesWritten: Int64,
+        totalBytesWritten: Int64,
+        totalBytesExpectedToWrite: Int64
+    ) {
+        let download = self.download(for: downloadTask)
+        let progress = Double(totalBytesWritten) / Double(totalBytesExpectedToWrite)
+        download.progress(progress)
+    }
+
+    public func urlSession(
+        _ session: URLSession,
+        downloadTask: URLSessionDownloadTask,
+        didFinishDownloadingTo location: URL
+    ) {
+        let download = self.download(for: downloadTask)
+        let response = downloadTask.response as! HTTPURLResponse
+
+        guard (200..<300).contains(response.statusCode) else {
+            download.completion(.failure(DownloaderError.serverError(statusCode: response.statusCode)))
+            return
+        }
+
+        let path = AbsolutePath(location.path)
+        download.completion(.success(path))
+    }
+
+    public func urlSession(_ session: URLSession, task: URLSessionTask, didCompleteWithError error: Error?) {
+        if let error = error {
+            let download = self.download(for: task)
+            download.completion(.failure(DownloaderError.clientError(error: error)))
+        }
+    }
+}
+
+extension FoundationDownloader {
+
+    /// Returns the download data structure associated with a task and fails if none is found as this should never
+    /// happen.
+    private func download(for task: URLSessionTask) -> Download {
+        guard let download = downloads[task.taskIdentifier] else {
```

Comment:
I think we should make this thread-safe

Reply:
It’s always called from functions on the same queue, so it’s not really necessary. What’s the reasoning?

Reply:
This is called in all URLSessionDownloadDelegate methods which could be running in parallel when multiple tasks are added, right?

Reply:
Explained why its okay in the comments.